clustered/client/src/main/java/org/ehcache/clustered/client/internal/config/xml/ClusteringCacheManagerServiceConfigurationParser.java

```diff
@@ -0,0 +1,466 @@
+    rootElement.appendChild(serverSideConfigurationElem);
+    return rootElement;
+  }
+
+  private void validateParametersForTranslationToServiceConfig(Document doc
+    , ServiceCreationConfiguration<ClusteringService> serviceCreationConfiguration) {
+    Objects.requireNonNull(doc, "Document must not be NULL");
+    Objects.requireNonNull(serviceCreationConfiguration, "ServiceCreationConfiguration must not be NULL");
+    if (!(serviceCreationConfiguration instanceof ClusteringServiceConfiguration)) {
+      throw new IllegalArgumentException("Parameter serviceCreationConfiguration must be of type ClusteringServiceConfiguration."
+                                         + "Provided type of parameter is : " + serviceCreationConfiguration.getClass());
+    }
+  }
+
+  private Element createRootUrlElement(Document doc, ClusteringServiceConfiguration clusteringServiceConfiguration) {
+    Element rootElement = doc.createElement(TC_CLUSTERED_NAMESPACE_PREFIX + COLON + CLUSTER_ELEMENT_NAME);
+    rootElement.setAttributeNS(XMLConstants.XMLNS_ATTRIBUTE_NS_URI, "xmlns:" + TC_CLUSTERED_NAMESPACE_PREFIX, NAMESPACE.toString());
+    Element urlElement = doc.createElement(TC_CLUSTERED_NAMESPACE_PREFIX + COLON + CONNECTION_ELEMENT_NAME);
+    urlElement.setAttribute(URL_ATTRIBUTE_NAME, clusteringServiceConfiguration.getClusterUri().toString());
+    rootElement.appendChild(urlElement);
+    return rootElement;
+  }
+
+  private Element createServerElement(Document doc, ClusteringServiceConfiguration clusteringServiceConfiguration) {
+    if (!(clusteringServiceConfiguration.getConnectionSource() instanceof ConnectionSource.ServerList)) {
+      throw new IllegalArgumentException("When connection URL is null, source of connection MUST be of type ConnectionSource.ServerList.class");
+    }
+    ConnectionSource.ServerList servers = (ConnectionSource.ServerList)clusteringServiceConfiguration.getConnectionSource();
+    if (!servers.getServers().iterator().hasNext()) {
+      throw new IllegalArgumentException("Empty server lists");
+    }
+    Element rootElement = doc.createElement(TC_CLUSTERED_NAMESPACE_PREFIX + COLON + CLUSTER_ELEMENT_NAME);
+    rootElement.setAttributeNS(XMLConstants.XMLNS_ATTRIBUTE_NS_URI, "xmlns:" + TC_CLUSTERED_NAMESPACE_PREFIX, NAMESPACE.toString());
+    Element connElement = doc.createElement(TC_CLUSTERED_NAMESPACE_PREFIX + COLON + CLUSTER_CONNECTION_ELEMENT_NAME);
+    connElement.setAttribute(CLUSTER_TIER_MANAGER_ATTRIBUTE_NAME, clusteringServiceConfiguration.getConnectionSource()
+      .getClusterTierManager());
+    servers.getServers().forEach(server -> {
+      Element serverElement = doc.createElement(TC_CLUSTERED_NAMESPACE_PREFIX + COLON + SERVER_ELEMENT_NAME);
+      serverElement.setAttribute(HOST_ATTRIBUTE_NAME, server.getHostName());
+      if (server.getPort() < 0) {
+        throw new IllegalArgumentException("Port for server:" + server.getHostName() + " must not be negative");
```

Comment:
The schema will validate this for us won't it?

Reply:
I wanted to make explicit check so that I can throw an exception with custom message, instead a generic parser validation error. Let me know if I should depend only on schema validator.

Reply:
I would always lean toward having these kinds of rules in as few places as possible, then lets see what the errors look like... if they are to impenetrable we can make a decision then.